### ISSUE TITLE
deal with unfinalized blocks

### ISSUE THREAD

**Issue by shunjizhan** [AcalaNetwork/bodhi.js#120]
as discussed here https://github.com/AcalaNetwork/bodhi.js/issues/73#issuecomment-964798469, we need a helper JS class to listen to recent/finalized block change, and index these unfinalized block data.

**Comment by xlc** [comment]
Something like this. This allow constant time add and purge block data 

```
interface BlockData {
  blockNumber: number
  bucket: number
  block: Block
  transacrtions: Record<string, Tranasction> // txhas => tx
}

const BUCKET_SIZE = 5

export class UnfinalizedBlockCache {
  buckets: Array<Array<BlockData>> = []

  addBlock(block: Block) {
    const bucket = Math.floor(block.number / BUCKET_SIZE)
    const startBucket = this.buckets[0]?.[0]?.bucket ?? 0
    const idx = bucket - startBucket
    this.buckets[idx] = this.buckets[idx] ?? []
    const bucketData = {
     bucket
     // other fields
    }
    this.buckets[idx][block.number % BUCKET_SIZE] = bucketData
  }

  updateFinalizedHead(blockNumber: number) {
    const bucket = Math.floor(blockNumber / BUCKET_SIZE)
    const startBucket = this.buckets[0]?.[0]?.bucket
    if (startBucket != null && bucket > startBucket) {
      const idx = bucket - startBucket
      this.buckets.splice(0, idx + 1) // make sure this can handle holes
    }
  }

  queryTransactionByHash(hash: string) {
    for (const i = this.buckets.length - 1; i >= 0; --i ) {
      const bucketData = this.buckets[i]  // search from latest bucket
      for (const blockData of bucketData) {
        const tx = blockData.transactions[hash]
        if (tx) {
          return tx
        }
      }
    }
  }
}

function createUnfinalizedBlockCache(api: ApiPromise) {
  const cache = new UnfinalizedBlockCache()
  let lastProcessedBlock = undefined

  // subscribe to new head
  // retry on error
    // while lastProcessedBlock !== undefiled && lastProcessedBlock < latestBlockNumber
      // let block = getBlock(lastProcessedBlock + 1)
      // cache.addBlock(block)
      // lastProcessedBlock = block.number

  // subscribe to finalized haed
  // retry on error
    // lastProcessedBlock = max(lastProcessedBlock, blockNumber)
    // cache.updateFinalizedHead(blockNumber)
  return cache
}

```

**Comment by shunjizhan** [comment]
closes with #131
